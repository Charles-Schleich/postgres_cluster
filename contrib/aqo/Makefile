# contrib/aqo/Makefile

MODULE_big = aqo
OBJS = aqo.o auto_tuning.o cardinality_estimation.o cardinality_hooks.o \
hash.o machine_learning.o path_utils.o postprocessing.o preprocessing.o \
selectivity_cache.o storage.o utils.o $(WIN32RES)
EXTENSION = aqo
DATA = aqo--1.0.sql
PGFILEDESC = "aqo - adaptive query optimization"

ifdef USE_PGXS
PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
else
subdir = contrib/aqo
top_builddir = ../..
include $(top_builddir)/src/Makefile.global
include $(top_srcdir)/contrib/contrib-global.mk
endif
