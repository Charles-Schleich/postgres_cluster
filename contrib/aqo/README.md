# Adaptive query optimization

Adaptive query optimization is the extension of standard PostgreSQL cost-based
query optimizer. Its basical principle is to use query execution statistics
for improving cardinality estimation. Experimental evaluation shows that this
improvement sometimes provides an enourmously large speed-up for rather
complicated queries.

## Installation

In your db:
CREATE EXTENSION aqo;

and modify your postgresql.conf:
shared_preload_libraries = 'aqo.so'

It is essential that library is preloaded during server startup, because
adaptive query optimization has to be enabled on per-database basis instead
of per-connection.

## Usage

Note that the extension works bad with dynamically generated views. If they
appear in workload, please use "aqo.mode='manual'".

This extension has intelligent self-tuning mode. If you want to rely completely
on it, just add line "aqo.mode = 'intelligent'" into your postgresql.conf.

Now this mode may work not good for rapidly changing data and query
distributions, so it is better to reset extension manually when that happens.

Also please note that intelligent mode is not supposed to work with queries
with dynamically generated structure. Dynamically generated constants are okay.

For handling workloads with dynamically generated query structures the forced
mode "aqo.mode = 'forced'" is provided. We cannot guarantee performance
improvement with this mode, but you may try it nevertheless.

If you want to completelly control how PostgreSQL optimizes queries, use manual
mode "aqo.mode = 'manual'" and
contrib/aqo/learn_queries.sh file_with_sql_queries.sql "psql -d YOUR_DATABASE"
where file_with_sql_queries.sql is a textfile with queries on which aqo is
supposed to learn. Please use only SELECT queries file_with_sql_queries.sql.
More sophisticated and convenient tool for aqo administration is in the
development now.
If you want to freeze optimizer's behaviour (i. e. disable learning under
workload), use "UPDATE aqo_queries SET auto_tuning=false;".
If you want to disable aqo for all queries, you may use
"UPDATE aqo_queries SET use_aqo=false, learn_aqo=false, auto_tuning=false;".

## Advanced tuning

To control query optimization we introduce for each query its type.
We consider that queries belong to the same type if and only if they differ only
in their constants.
One can see an example of query corresponding to the specified query type
in table aqo_query_texts.
select * from aqo_query_texts;

That is why intelligent mode does not work for dynamically generated query
structures: it tries to learn separately how to optimize different query types,
and for dynamical query structure the query types are different, so it will
consume a lot of memory and will not optimize any query properly.

Forced mode forces aqo to ignore query types and optimize them together. On one
hand it lacks of intelligent tuning, so the performance for some queries may
even decrease, on the other hand it may work for dynamic workload and consumes
less memory than the intellignet mode. That is why you may want to use it.

Each query type has its own optimization settings. You can find them in table
aqo_queries.

Auto_tuning setting identifies whether aqo module tries to tune other settings
from aqo_queries for the query type. If the mode is intelligent, defalt value
for new queries is true. If the mode is not intelligent, new queries are not
appended to aqo_queries automatically, but you can also set auto_tuning variable
to true manually.

Use_aqo setting shows whether aqo cardinalities prediction be used for next
execution of such query type. Disabling of aqo usage is reasonable for that
cases in which query execution time increases after applying aqo. It happens
sometimes because of cost models incompleteness.

Learn_aqo setting shows whether aqo collects statistics for next execution of
such query type. True value may have computational overheads, but it is
essential when aqo model does not fit the data. It happens at the start of aqo
for the new query type or when the data distribution in database is changed.

Fspace_hash setting is for extra advanced aqo tuning. It may be changed manually
to optimize a number of query types using the same model. It may decrease the
amount of memory for models and even query execution performance, but also it
may cause the bad aqo's behaviour, so please use it only if you know exactly
what you do.

## Statistics

For forced and intelligent query modes, and for all tracked queries the
statistics is collected. The statistics is cardinality quality, planning and
execution time. For forced mode the statistics for all untracked query types
is stored in common query type with hash 0.

One can see the collected statistics in table aqo_query_stat.
