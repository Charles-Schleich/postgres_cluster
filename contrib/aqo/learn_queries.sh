#!/bin/bash

(>&2 echo "Started.")
((echo "BEGIN TRANSACTION;") && (echo "SET aqo.mode='intelligent';") && (for ((i = 0; i < ${3:-20}; ++i)); do cat $1 && (echo ";"); done) && (echo "COMMIT TRANSACTION;")) | $2 >aqo_learning_log.out 2>aqo_learning_log.err
(>&2 echo "Finished.")
