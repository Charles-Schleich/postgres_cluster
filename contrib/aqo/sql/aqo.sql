CREATE TABLE aqo_test0(a int, b int, c int, d int);
INSERT INTO aqo_test0 VALUES (0, 0, 0, 0);
INSERT INTO aqo_test0 VALUES (1, 1, 1, 1);
INSERT INTO aqo_test0 VALUES (3, 3, 3, 3);
INSERT INTO aqo_test0 VALUES (5, 5, 5, 5);
INSERT INTO aqo_test0 VALUES (6, 6, 6, 6);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
ANALYZE aqo_test0;

CREATE EXTENSION aqo;

SET aqo.mode = 'manual';

EXPLAIN 
SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;

EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;

SET aqo.mode = 'forced';

CREATE TABLE tmp1 AS SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;
SELECT count(*) FROM tmp1;
DROP TABLE tmp1;

CREATE TABLE tmp1 AS SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;
SELECT count(*) FROM tmp1;
DROP TABLE tmp1;

EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;

EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;

DROP TABLE aqo_test0;

DROP EXTENSION aqo;
