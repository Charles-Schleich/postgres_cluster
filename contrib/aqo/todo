Roadmap:
1. Make learn query utility.
2. Make intelligent mode more intelligent.
3. Further research.

Todo, ideas:
1a. set_baserel_size_estimates needs to be splitted for the beauty of code
1b. set_baserel_rows_estimates sounds not good
2. More pain: call_default MACROS
3. Avoid memory leaks.
4. clausesel.c orclause case to be fixed
5. Where to write about hook on copy_generic_path_info?
6. Do we need initPlan in collect_planstat?
7. Problems with pfree when aqo_K changed during operation (impossbile?)
8. GUC for ml parameters?
9. object_max_distance (prediction refusal)?
10. Append and MergeAppend paths - what with them?
	ResultsPath MinMaxAgrrPath RecursiveUnionPath
11. ModifyTablePath!
12. Better to store floats, not doubles.
13. Intelligence mode is better to have optimization sandbox.
14. For intelligent mode: provide a sandbox to train unstable query types.
15. For intelligent mode: provide a possibility (and mechanism) for running
	queries on ours own wish for learning purposes.
16. Views with the same names.
17. Maybe hotfix for completely text-equal queries? What with constants
	in this case?

Code sketches:
#include "utils/lsyscache.h"
int16 typlen;
bool typbyval;
char typalign;
get_typlenbyvalalign(FLOAT8OID, &typlen, &typbyval, &typalign);
elog(WARNING, "%d %d %c", (int) typlen, (int) typbyval, typalign);
