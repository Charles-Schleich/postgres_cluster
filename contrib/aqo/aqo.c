#include "aqo.h"

PG_MODULE_MAGIC;

void		_PG_init(void);
void		_PG_fini(void);

/* Strategy of determining feature space for new queries. */
int			aqo_mode;

/* GUC variables */
static const struct config_enum_entry format_options[] = {
	{"intelligent", AQO_MODE_INTELLIGENT, false},
	{"forced", AQO_MODE_FORCED, false},
	{"controlled", AQO_MODE_CONTROLLED, false},
	{"disabled", AQO_MODE_DISABLED, false},
	{NULL, 0, false}
};

/* Parameters of autotuning */
int			aqo_stat_size = 20;
int			auto_tuning_window_size = 5;
double		auto_tuning_exploration = 0.1;
int			auto_tuning_max_iterations = 50;
int			auto_tuning_infinite_loop = 8;

/* stat_size > infinite_loop + window_size + 3 is required for auto_tuning*/

/* Machine learning parameters */
double		object_selection_prediction_threshold = 0.3;
double		object_selection_object_threshold = 0.1;
double		learning_rate = 1e-1;
int			aqo_k = 3;
int			aqo_K = 30;
double		log_selectivity_lower_bound = -30;

/* Parameters for current query */
int			query_hash;
bool		learn_aqo;
bool		use_aqo;
int			fspace_hash;
bool		auto_tuning;
bool		collect_stat;
bool		adding_query;
bool		explain_only;
bool		explain_aqo;

/* Query execution time */
instr_time	query_starttime;
double		query_planning_time;

/* Saved hook values in case of unload */
post_parse_analyze_hook_type prev_post_parse_analyze_hook;
planner_hook_type prev_planner_hook;
ExecutorStart_hook_type prev_ExecutorStart_hook;
ExecutorEnd_hook_type prev_ExecutorEnd_hook;
set_baserel_rows_estimate_hook_type prev_set_baserel_rows_estimate_hook;
get_parameterized_baserel_size_hook_type prev_get_parameterized_baserel_size_hook;
set_joinrel_size_estimates_hook_type prev_set_joinrel_size_estimates_hook;
get_parameterized_joinrel_size_hook_type prev_get_parameterized_joinrel_size_hook;
copy_generic_path_info_hook_type prev_copy_generic_path_info_hook;
ExplainOnePlan_hook_type prev_ExplainOnePlan_hook;

/*****************************************************************************
 *
 *	CREATE/DROP EXTENSION FUNCTIONS
 *
 *****************************************************************************/

void
_PG_init(void)
{
	DefineCustomEnumVariable("aqo.mode",
							 "Mode of aqo usage.",
							 NULL,
							 &aqo_mode,
							 AQO_MODE_CONTROLLED,
							 format_options,
							 PGC_SUSET,
							 0,
							 NULL,
							 NULL,
							 NULL);

	prev_planner_hook = planner_hook;
	planner_hook = &aqo_planner;
	prev_post_parse_analyze_hook = post_parse_analyze_hook;
	post_parse_analyze_hook = &get_query_text;
	prev_ExecutorStart_hook = ExecutorStart_hook;
	ExecutorStart_hook = &aqo_ExecutorStart;
	prev_ExecutorEnd_hook = ExecutorEnd_hook;
	ExecutorEnd_hook = &learn_query_stat;
	prev_set_baserel_rows_estimate_hook = set_baserel_rows_estimate_hook;
	set_baserel_rows_estimate_hook = &aqo_set_baserel_rows_estimate;
	prev_get_parameterized_baserel_size_hook =
		get_parameterized_baserel_size_hook;
	get_parameterized_baserel_size_hook =
		&aqo_get_parameterized_baserel_size;
	prev_set_joinrel_size_estimates_hook = set_joinrel_size_estimates_hook;
	set_joinrel_size_estimates_hook = &aqo_set_joinrel_size_estimates;
	prev_get_parameterized_joinrel_size_hook =
		get_parameterized_joinrel_size_hook;
	get_parameterized_joinrel_size_hook =
		&aqo_get_parameterized_joinrel_size;
	prev_copy_generic_path_info_hook = copy_generic_path_info_hook;
	copy_generic_path_info_hook = &aqo_copy_generic_path_info;
	prev_ExplainOnePlan_hook = ExplainOnePlan_hook;
	ExplainOnePlan_hook = print_into_explain;
	init_deactivated_queries_storage();
}

void
_PG_fini(void)
{
	planner_hook = prev_planner_hook;
	post_parse_analyze_hook = prev_post_parse_analyze_hook;
	ExecutorStart_hook = prev_ExecutorStart_hook;
	ExecutorEnd_hook = prev_ExecutorEnd_hook;
	set_baserel_rows_estimate_hook = prev_set_baserel_rows_estimate_hook;
	get_parameterized_baserel_size_hook =
		prev_get_parameterized_baserel_size_hook;
	set_joinrel_size_estimates_hook = prev_set_joinrel_size_estimates_hook;
	get_parameterized_joinrel_size_hook =
		prev_get_parameterized_joinrel_size_hook;
	copy_generic_path_info_hook = prev_copy_generic_path_info_hook;
	ExplainOnePlan_hook = prev_ExplainOnePlan_hook;
	fini_deactivated_queries_storage();
}

PG_FUNCTION_INFO_V1(invalidate_deactivated_queries_cache);

/*
 * Clears the cache of deactivated queries if the user changed aqo_queries
 * manually.
 */
Datum
invalidate_deactivated_queries_cache(PG_FUNCTION_ARGS)
{
	fini_deactivated_queries_storage();
	init_deactivated_queries_storage();
	PG_RETURN_POINTER(NULL);
}
