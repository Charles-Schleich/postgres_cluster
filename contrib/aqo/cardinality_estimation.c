#include "aqo.h"

/*****************************************************************************
 *
 *	CARDINALITY ESTIMATION
 *
 * This is the module in which cardinality estimation problem obtained from
 * cardinality_hooks turns into machine learning problem.
 *
 *****************************************************************************/

/*
 * General method for prediction the cardinality of given relation.
 */
double
predict_for_relation(List *restrict_clauses, List *selectivities, List *relids)
{
	int			nfeatures;
	int			fss_hash;
	double	  **matrix;
	double	   *target;
	double	   *features;
	double		result;
	int			rows;
	int			i;
	ListCell   *l;

	get_fss_for_object(restrict_clauses, selectivities, relids,
					   &nfeatures, &fss_hash, &features);

	matrix = palloc(sizeof(*matrix) * aqo_K);
	for (i = 0; i < aqo_K; ++i)
		matrix[i] = palloc0(sizeof(**matrix) * nfeatures);
	target = palloc0(sizeof(*target) * aqo_K);

	if (load_fss(fss_hash, nfeatures, matrix, target, &rows))
		result = OkNNr_predict(rows, nfeatures, matrix, target, features);
	else
		result = -1;

	pfree(features);
	for (i = 0; i < aqo_K; ++i)
		pfree(matrix[i]);
	pfree(matrix);
	pfree(target);
	list_free(restrict_clauses);
	foreach(l, selectivities)
		pfree(lfirst(l));
	list_free(selectivities);
	list_free(relids);

	if (result < 0)
		return -1;
	else
		return exp(result);
}
