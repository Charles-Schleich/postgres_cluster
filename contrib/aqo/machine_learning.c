#include "aqo.h"

/*****************************************************************************
 *
 *	MACHINE LEARNING TECHNIQUES
 *
 * This module doesn't know anything about DBMS, cardinalities and all other
 * stuff. It learns matrixes, predicts values and is quite happy.
 * The proposed method is designed for working with limited number of objects.
 * It is guaranteed that number of rows in the matrix will not exceed aqo_K
 * setting after learning procedure. This property also allows to adapt to
 * workloads which properties are slowly changed.
 *
 *****************************************************************************/

static double fs_distance(double *a, double *b, int len);
static double fs_similarity(double dist);
static void OkNNr_compute_weights(double *distances, int matrix_rows,
					  double **w, double *w_sum, int **idx);


/*
 * Computes L2-distance between two given vectors.
 */
double
fs_distance(double *a, double *b, int len)
{
	double		res = 0;
	int			i;

	for (i = 0; i < len; ++i)
		res += (a[i] - b[i]) * (a[i] - b[i]);
	if (len != 0)
		res = sqrt(res / len);
	return res;
}

/*
 * Returns similarity between objects based on distance between them.
 */
double
fs_similarity(double dist)
{
	return 1.0 / (0.1 + dist);
}

/*
 * Compute weights necessary for both prediction and learning.
 * Creates and returns w, w_sum and idx based on given distances ad matrix_rows.
 *
 * Appeared as a separate function because of "don't repeat your code"
 * principle.
 */
void
OkNNr_compute_weights(double *distances, int matrix_rows,
					  double **w, double *w_sum, int **idx)
{
	int			i,
				j;
	int			to_insert,
				tmp;

	*w_sum = 0;

	*idx = palloc0(sizeof(**idx) * aqo_k);
	for (i = 0; i < aqo_k; ++i)
		(*idx)[i] = -1;
	(*w) = palloc0(sizeof(**w) * aqo_k);

	for (i = 0; i < matrix_rows; ++i)
		for (j = 0; j < aqo_k; ++j)
			if ((*idx)[j] == -1 || distances[i] < distances[(*idx)[j]])
			{
				to_insert = i;
				for (; j < aqo_k; ++j)
				{
					tmp = (*idx)[j];
					(*idx)[j] = to_insert;
					to_insert = tmp;
				}
				break;
			}
	for (i = 0; i < aqo_k && (*idx)[i] != -1; ++i)
	{
		(*w)[i] = fs_similarity(distances[(*idx)[i]]);
		*w_sum += (*w)[i];
	}
}

/*
 * With given matrix, targets and features makes prediction for current object.
 *
 * Returns negative value in the case of refusal to make a prediction, because
 * positive targets are assumed.
 */
double
OkNNr_predict(int matrix_rows, int matrix_cols,
			  double **matrix, double *targets,
			  double *nw_features)
{
	double	   *distances;
	int			i;
	int		   *idx;
	double	   *w;
	double		w_sum;
	double		result = 0;

	distances = palloc0(sizeof(*distances) * matrix_rows);

	for (i = 0; i < matrix_rows; ++i)
		distances[i] = fs_distance(matrix[i], nw_features, matrix_cols);

	OkNNr_compute_weights(distances, matrix_rows, &w, &w_sum, &idx);

	for (i = 0; i < aqo_k; ++i)
		if (idx[i] != -1)
			result += targets[idx[i]] * w[i] / w_sum;

	pfree(distances);
	pfree(idx);
	pfree(w);

	/* this should never happen */
	if (idx[0] == -1)
		return -1;

	if (result < 0)
		result = 0;
	return result;
}

/*
 * Modifies given matrix and targets using features and target value of new
 * object.
 * Returns indexes of changed lines: if index of line is less than matrix_rows
 * updates this line in database, otherwise adds new line with given index.
 * It is supposed that indexes of new lines are consequent numbers
 * starting from matrix_rows.
 */
List *
OkNNr_learn(int matrix_rows, int matrix_cols,
			double **matrix, double *targets,
			double *nw_features, double nw_target)
{
	List	   *lst = NIL;
	double	   *distances;
	int			i,
				j;
	int			min_distance_id = 0;
	int		   *idx;
	double	   *w;
	double		w_sum;
	double	   *cur_row;
	double		coef1,
				coef2;
	double		result = 0;

	distances = palloc0(sizeof(*distances) * matrix_rows);

	for (i = 0; i < matrix_rows; ++i)
	{
		distances[i] = fs_distance(matrix[i], nw_features, matrix_cols);
		if (distances[i] < distances[min_distance_id])
			min_distance_id = i;
	}
	if (matrix_rows < aqo_K)
	{
		if (matrix_rows != 0 && distances[min_distance_id] <
			object_selection_object_threshold)
		{
			for (j = 0; j < matrix_cols; ++j)
				matrix[min_distance_id][j] += learning_rate *
					(nw_features[j] - matrix[min_distance_id][j]);
			targets[min_distance_id] += learning_rate *
				(nw_target - targets[min_distance_id]);
			lst = lappend_int(lst, min_distance_id);
		}
		else
		{
			for (j = 0; j < matrix_cols; ++j)
				matrix[matrix_rows][j] = nw_features[j];
			targets[matrix_rows] = nw_target;
			lst = lappend_int(lst, matrix_rows);
		}
	}
	else
	{
		OkNNr_compute_weights(distances, matrix_rows, &w, &w_sum, &idx);

		for (i = 0; i < aqo_k && idx[i] != -1; ++i)
			result += targets[idx[i]] * w[i] / w_sum;
		coef1 = learning_rate * (result - nw_target);

		for (i = 0; i < aqo_k && idx[i] != -1; ++i)
		{
			coef2 = coef1 * (targets[idx[i]] - result) * w[i] * w[i] /
				sqrt(matrix_cols) / w_sum;

			targets[idx[i]] -= coef1 * w[i] / w_sum;
			for (j = 0; j < matrix_cols; ++j)
			{
				cur_row = matrix[idx[i]];
				cur_row[j] -= coef2 * (nw_features[j] - cur_row[j]) /
					distances[idx[i]];
			}

			lst = lappend_int(lst, idx[i]);
		}

		pfree(w);
		pfree(idx);
	}

	pfree(distances);
	return lst;
}
