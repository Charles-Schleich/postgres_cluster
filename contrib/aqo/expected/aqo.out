CREATE TABLE aqo_test0(a int, b int, c int, d int);
INSERT INTO aqo_test0 VALUES (0, 0, 0, 0);
INSERT INTO aqo_test0 VALUES (1, 1, 1, 1);
INSERT INTO aqo_test0 VALUES (3, 3, 3, 3);
INSERT INTO aqo_test0 VALUES (5, 5, 5, 5);
INSERT INTO aqo_test0 VALUES (6, 6, 6, 6);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
INSERT INTO aqo_test0 (SELECT * FROM aqo_test0);
ANALYZE aqo_test0;
CREATE EXTENSION aqo;
SET aqo.mode = 'manual';
EXPLAIN 
SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;
                          QUERY PLAN                          
--------------------------------------------------------------
 Seq Scan on aqo_test0  (cost=0.00..130.40 rows=131 width=16)
   Filter: ((a < 3) AND (b < 3) AND (c < 3) AND (d < 3))
(2 rows)

EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;
                          QUERY PLAN                          
--------------------------------------------------------------
 Seq Scan on aqo_test0  (cost=0.00..130.40 rows=664 width=16)
   Filter: ((a < 5) AND (b < 5) AND (c < 5) AND (d < 5))
(2 rows)

SET aqo.mode = 'forced';
CREATE TABLE tmp1 AS SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;
SELECT count(*) FROM tmp1;
 count 
-------
  2048
(1 row)

DROP TABLE tmp1;
CREATE TABLE tmp1 AS SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;
SELECT count(*) FROM tmp1;
 count 
-------
  3072
(1 row)

DROP TABLE tmp1;
EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 3 AND b < 3 AND c < 3 AND d < 3;
                          QUERY PLAN                           
---------------------------------------------------------------
 Seq Scan on aqo_test0  (cost=0.00..130.40 rows=2050 width=16)
   Filter: ((a < 3) AND (b < 3) AND (c < 3) AND (d < 3))
(2 rows)

EXPLAIN
SELECT * FROM aqo_test0
WHERE a < 5 AND b < 5 AND c < 5 AND d < 5;
                          QUERY PLAN                           
---------------------------------------------------------------
 Seq Scan on aqo_test0  (cost=0.00..130.40 rows=3069 width=16)
   Filter: ((a < 5) AND (b < 5) AND (c < 5) AND (d < 5))
(2 rows)

DROP TABLE aqo_test0;
DROP EXTENSION aqo;
